src/thread.rs

```diff
@@ -578,27 +658,52 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 
     #[inline]
-    fn create_blockset(&mut self) -> InterpResult<'tcx, BlockSetId> {
+    fn block_thread(&mut self, thread: ThreadId) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        Ok(this.machine.threads.block_thread(thread))
+    }
+
+    #[inline]
+    fn unblock_thread(&mut self, thread: ThreadId) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        Ok(this.machine.threads.unblock_thread(thread))
+    }
+
+    #[inline]
+    fn yield_active_thread(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        Ok(this.machine.threads.create_blockset())
+        this.machine.threads.yield_active_thread();
+        Ok(())
     }
 
     #[inline]
-    fn block_active_thread(&mut self, set: BlockSetId) -> InterpResult<'tcx> {
+    fn register_timeout_callback(
+        &mut self,
+        thread: ThreadId,
+        call_time: Time,
+        callback: TimeoutCallback<'mir, 'tcx>,
+    ) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        Ok(this.machine.threads.block_active_thread(set))
+        this.machine.threads.register_timeout_callback(thread, call_time, callback);
+        Ok(())
     }
 
     #[inline]
-    fn unblock_some_thread(&mut self, set: BlockSetId) -> InterpResult<'tcx, Option<ThreadId>> {
+    fn unregister_timeout_callback_if_exists(&mut self, thread: ThreadId) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        Ok(this.machine.threads.unblock_some_thread(set))
+        this.machine.threads.unregister_timeout_callback_if_exists(thread);
+        Ok(())
     }
 
+    /// Execute a timeout callback on the callback's thread.
     #[inline]
-    fn yield_active_thread(&mut self) -> InterpResult<'tcx> {
+    fn run_timeout_callback(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        this.machine.threads.yield_active_thread();
+        let (thread, callback) =
+            this.machine.threads.get_ready_callback().expect("no callback found");
+        let old_thread = this.set_active_thread(thread)?;
+        callback(this)?;
+        this.set_active_thread(old_thread)?;
```

Comment:
This back-and-forth of the active thread is curious, why do we have to do this?

Reply:
I guess an alternative design would be to not associate callbacks with threads at all. Then it would be up to the callback to work no matter what the active thread is. Not sure if that's too error-prone though?

Reply:
> This back-and-forth of the active thread is curious, why do we have to do this?

It was simply my design decision to make the caller and not the callback responsible for activating the correct thread. I set the active thread back to the original to make the scheduler the only real point that changes the thread being executed (just a design decision, I do not have a strong opinion on this).

>  Looks like this makes it impossible for the callback to activate the thread it is associated with?

Yes, that is a limitation. However, do we have a use-case for this?

> I guess an alternative design would be to not associate callbacks with threads at all. Then it would be up to the callback to work no matter what the active thread is. Not sure if that's too error-prone though?

Currently, the callbacks are associated with threads so that a thread can cancel its callback. Therefore, I am not sure if we can completely avoid associating callbacks with threads.

Reply:
> It was simply my design decision to make the caller and not the callback responsible for activating the correct thread. I set the active thread back to the original to make the scheduler the only real point that changes the thread being executed (just a design decision, I do not have a strong opinion on this).

That's fair, as long as it is thoroughly documented.